packages/devextreme/js/__internal/ui/collection/m_collection_widget.edit.ts

```diff
@@ -467,12 +473,37 @@ const CollectionWidget = BaseCollectionWidget.inherit({
     }
   },
 
+  _updateSelectedItems(args) {
+    this._selection.onSelectionChanging();
+    if (this.isSelectionCancel instanceof Promise) {
+      this.isSelectionCancel.then((cancel) => {
+        if (!cancel) {
+          this.updatingSelectedItems(args);
+        } else {
+          this._setOptionWithoutOptionChange('selectedIndex', this.initialSelectedIndex);
```

Comment:
[excess code, difficult logic, performance, possible drawbacks]
So, how it works now as far as i see:
- selection is updated in the Selection module => we get notification about it through [onSelectionChanged handler](https://github.com/DevExpress/DevExtreme/pull/27860/files#diff-b8238af01aa4298df01c6c2d89d13db014bd657a316cb3bf9012c19da0dfcd86L135)
- we immediately update the Collection selectedKeys in the [onSelectionChanged](https://github.com/DevExpress/DevExtreme/pull/27860/files#diff-b8238af01aa4298df01c6c2d89d13db014bd657a316cb3bf9012c19da0dfcd86R141) handler but [store initial selectedIndex](https://github.com/DevExpress/DevExtreme/pull/27860/files#diff-b8238af01aa4298df01c6c2d89d13db014bd657a316cb3bf9012c19da0dfcd86R410) to the private variable
- then in this function we restore selection if `cancel` is true

This logic is very difficult to be understood, it definitely takes additional execution time (because we change selectedKeys and then restore it if selection was cancelled) and mostly has side effects, particularly it can not work with multiple selection.

So I suggest you reconsider the solution and try to get rid of selection restoring. Maybe for this it's enough to move  [this row](https://github.com/DevExpress/DevExtreme/pull/27860/files#diff-b8238af01aa4298df01c6c2d89d13db014bd657a316cb3bf9012c19da0dfcd86L137) to the `updatingSelectedItems ` - not sure, should be checked.

Reply:
[b62f308](https://github.com/DevExpress/DevExtreme/pull/27860/commits/b62f30857d83924a9f3c3b9e14c53016e7d67bdb)